KAFKA-16314: Introducing the AbortableTransactionException

sjhajharia opened the pull request:
As a part of KIP-890, we are introducing a new class of Exceptions which when encountered shall lead to Aborting the ongoing Transaction. The following PR introduces the same with client side handling and server side changes.

- On client Side, the code attempts to handle the exception as an Abortable error and ensure that it doesn't take the producer to a fatal state. For each of the Transactional APIs, we have added the appropriate handling. For the produce request, we have verified that the exception transitions the state to Aborted.
- On the server side, we have bumped the ProduceRequest, ProduceResponse, TxnOffestCommitRequest and TxnOffsetCommitResponse Version. The appropriate handling on the server side has been added to ensure that the new error case is sent back only for the new clients. The older clients will continue to get the old Invalid_txn_state exception to maintain backward compatibility.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

sjhajharia committed:
Client Sid handling of the Abortable Error

sjhajharia committed:
Server side changes for handling the AbortableTxnsException

jolshan commented on clients/src/main/java/org/apache/kafka/common/errors/AbortableTransactionException.java:
```diff
@@ -0,0 +1,23 @@
+ */
+package org.apache.kafka.common.errors;
+
+public class AbortableTransactionException extends ApiException {
```
I know there were some questions about the naming -- this seems reasonable but just wanted to call it out.

cc: @artemlivshits

jolshan replied:
I'm still thinking on this name as well. Since all transactions are technically abortable, and what we want to convey is that the only way forward is to abort, would it make sense to say TransactionAbortableException?

Maybe it doesn't make a huge difference though. 😅 Wondering if there is a way to make that distinction of abortable vs we need to abort to proceed. I realize though that we are also modeling off of RetriableException here and trying to keep that similar.

AndrewJSchofield replied:
I tend to agree that `TransactionAbortableException` is very slightly more satisfying than `AbortableTransactionException`.

jolshan commented on clients/src/main/resources/common/message/ProduceRequest.json:
```diff
@@ -35,7 +35,10 @@
   // Version 10 is the same as version 9 (KIP-951).
-  "validVersions": "0-10",
+  //
+  // Version 11 is schematically same as Version 10. It will only enable the server to know that
```
For now, let's just say Version 11 is the same as 10. We may decide to do a separate bump for the implicit add partitions to transaction.

jolshan commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:
```diff
@@ -3151,6 +3152,45 @@ public void testInvalidTxnStateIsAnAbortableError() throws Exception {
+        sender.runOnce();  // send request
+        sendIdempotentProducerResponse(0, tp0, Errors.ABORTABLE_TRANSACTION_EXCEPTION, -1);
+
+        // Return InvalidTxnState error. It should be abortable.
```
Is this comment correct?

sjhajharia replied:
fixed

jolshan commented on core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala:
```diff
@@ -228,11 +232,17 @@ class AddPartitionsToTxnManager(
                       Errors.INVALID_PRODUCER_EPOCH.code
                     else
                       partitionResult.partitionErrorCode
+                  // For backward compatibility with clients.
```
nice 👍
 I think we can include this in val code as a subcase so we don't need the var

sjhajharia replied:
agreed. thanks for the suggestion
I will update the code here

jolshan commented on core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala:
```diff
@@ -49,7 +49,8 @@ object AddPartitionsToTxnManager {
                                   val callbacks: mutable.Map[String, AddPartitionsToTxnManager.AppendCallback],
-                                  val startTimeMs: mutable.Map[String, Long])
+                                  val startTimeMs: mutable.Map[String, Long],
+                                  val produceRequestVersion: Short)
```
nice this is what I was thinking 👍 
I noticed that this is a bit tricky for the offset commit path. We will want a given request version there too, but the versions don't map quite the same. I will think a bit on how to address this.

jolshan commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -1007,7 +1010,8 @@ class ReplicaManager(val config: KafkaConfig,
       producerEpoch,
-      generalizedCallback
+      generalizedCallback,
+      ApiKeys.PRODUCE.latestVersion
```
Hmm this is a bit interesting since this path is usually exercised by the group coordinator. We may want to do something else, or leave a comment here.

sjhajharia replied:
I don't think that including the latest Produce Request Version will break any code flow. So should we just leave a comment denoting the same?

jolshan replied:
What if we send a offset commit request that can't yet handle the error.
I think we may need a different data structure. Maybe something like enums to handle the various versions 
0 -- default, 1 -- new error, 2 -- add partitions rather than verifying.

sjhajharia replied:
So for this particular path it would be 0 (default)
And for the new workflow we added, it would be new error (1)
Where can we use the 3rd use case --> that is add partitions rather than verifying?

jolshan replied:
This will be done as a follow up and part of KIP-890 part 2
cc: @CalvinConfluent who was also thinking about this part

jolshan commented:
I took a quick skim of the tests -- just wanted to confirm we have a test for the old produce request + invalid txn state returned?

sjhajharia committed:
Addressed some initial comments

sjhajharia commented:
> I took a quick skim of the tests -- just wanted to confirm we have a test for the old produce request + invalid txn state returned?

@jolshan , if you have a look at `core/src/test/scala/unit/kafka/server/AddPartitionsToTxnManagerTest.scala` file, there is a test added around lines 381-390 wherein based on the Produce Request Version, we are testing if the returned error is `Invalid_txn_state` or `AbortableTransactionsException`

sjhajharia committed:
Fixed failing test

sjhajharia committed:
Adding enum

jolshan commented on core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala:
```diff
@@ -42,6 +42,11 @@ object AddPartitionsToTxnManager {
 
+object OperationExpected extends Enumeration {
+  type operation = Value
+  val defaultOperation, latestProduceVersion, addPartitionWithoutVerification = Value
```
I think we could have something like `default`, `genericError`, and `addPartition`?

jolshan replied:
I also think in scala we typically use sealed traits and case objects. 

See https://github.com/apache/kafka/blob/58ddd693e69599b177d09c2e384f31e7f5e11171/core/src/main/scala/kafka/controller/KafkaController.scala#L59

sjhajharia replied:
ack
updated the code to use `sealed trait`

sjhajharia committed:
Updated the code to use the sealed trait for Enum

jolshan commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -392,7 +393,8 @@ public enum Errors {
     TELEMETRY_TOO_LARGE(118, "Client sent a push telemetry request larger than the maximum size the broker will accept.", TelemetryTooLargeException::new),
-    INVALID_REGISTRATION(119, "The controller has considered the broker registration to be invalid.", InvalidRegistrationException::new);
+    INVALID_REGISTRATION(119, "The controller has considered the broker registration to be invalid.", InvalidRegistrationException::new),
+    ABORTABLE_TRANSACTION_EXCEPTION(120, "This maps to all errors which can be fixed by aborting the current transaction.", AbortableTransactionException::new);
```
nit: let's leave off the "exception" part to be consistent with the other errors here.

jolshan commented on core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala:
```diff
@@ -42,14 +42,26 @@ object AddPartitionsToTxnManager {
 
+/**
+ * This is an enum which handles the Partition Response based on the Produce Request Version and the exact operation
+ *    defaultOperation:   This is the default workflow which maps to cases when the Produce Request Version was lower than expected or when exercising the offset commit request path
```
I think we should include this change in the offset commit path. It will require bumping that request version as well. 

> We will bump the ProduceRequest/Response and TxnOffsetCommitRequest/Response version to indicate the client is using the new protocol that doesn’t require adding partitions to transactions and will implicitly do so. The bump will also support new errors ABORTABLE_ERROR

We are just doing 2 bumps for the above comment

jolshan replied:
It is a similar process to pass in the the value, but we will need it for both group coordinator paths. Let me know if you need help.

sjhajharia replied:
Done

sjhajharia committed:
Updated the Error name of Abortable Txn Error for consistency

CalvinLiu7947 commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java:
```diff
@@ -1504,6 +1507,8 @@ public void handleResponse(AbstractResponse response) {
                 fatalError(error.exception());
             } else if (error == Errors.GROUP_AUTHORIZATION_FAILED) {
                 abortableError(GroupAuthorizationException.forGroupId(key));
+            } else if (error == Errors.ABORTABLE_TRANSACTION_EXCEPTION) {
```
Does FindCoordinatorRequest also return the ABORTABLE_TRANSACTION_EXCEPTION?

jolshan replied:
The idea of this error is that we are future proofing. In the future if we want the producer to abort the transaction (say we have a new use case that requires this behavior) we can rely on this error to do the correct thing for older clients. (Or clients 3.8 and later in this case :) )

We ran into this issue a lot when picking error codes for kip-890 part 1. If we had such an error then, we could have used it for the old clients. Instead, we chosen INVALID_TXN_STATE which has inconsistent behavior across clients.

CalvinLiu7947 reviewed (commented):
Thanks for the change. 
Just have 1 question about the ABORTABLE_TRANSACTION_EXCEPTION. Will FindCoordinatorRequest and EndTxn requests also return ABORTABLE_TRANSACTION_EXCEPTION?

sjhajharia committed:
Bumping the Txn Offset Commit request and related changes

jolshan commented on core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala:
```diff
@@ -44,7 +44,7 @@ object AddPartitionsToTxnManager {
 /**
  * This is an enum which handles the Partition Response based on the Produce Request Version and the exact operation
- *    defaultOperation:   This is the default workflow which maps to cases when the Produce Request Version was lower than expected or when exercising the offset commit request path
+ *    defaultOperation:   This is the default workflow which maps to cases when the Produce Request Version or the Txn_offset_commit request was lower than expected
```
nit: replace "lower than expected" with "lower than the first version supporting the new error code"

jolshan commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java:
```diff
@@ -1456,7 +1456,7 @@ public <T> CompletableFuture<T> scheduleWriteOperation(
      * @param timeout           The write operation timeout.
      * @param op                The write operation.
-     *
+     * @param apiVersion        The Version of the Txn_Offset_Commit request
```
can we keep the new line below the last argument?

sjhajharia committed:
Merge upstream changes and resolve conflicts

sjhajharia committed:
Minor changes to comments

sjhajharia committed:
Fixed the comment for an updated request

jolshan commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -392,7 +393,8 @@ public enum Errors {
     TELEMETRY_TOO_LARGE(118, "Client sent a push telemetry request larger than the maximum size the broker will accept.", TelemetryTooLargeException::new),
-    INVALID_REGISTRATION(119, "The controller has considered the broker registration to be invalid.", InvalidRegistrationException::new);
+    INVALID_REGISTRATION(119, "The controller has considered the broker registration to be invalid.", InvalidRegistrationException::new),
+    ABORTABLE_TRANSACTION(120, "This maps to all errors which can be fixed by aborting the current transaction.", AbortableTransactionException::new);
```
Minor nit here with the comment. Maybe we can say -- something like
"The server encountered an error with the transaction. The client can abort the transaction to continue using this transactional ID."

jolshan commented:
I was thinking about this change and am wondering if we should bump all the transactional APIs to make it clear this error can be returned. Let me get back to you @sjhajharia -- I will consult some others.

jolshan commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:
```diff
@@ -3523,6 +3524,129 @@ public void testForegroundInvalidStateTransitionIsRecoverable() {
+        // make sure the exception was thrown directly from the follow-up calls.
+
+    }
+
```
nit: extra new line here.

jolshan commented on core/src/main/scala/kafka/server/ReplicaManager.scala:
```diff
@@ -975,12 +978,13 @@ class ReplicaManager(val config: KafkaConfig,
+   * @param producerEpoch       the epoch of the producer writing to the transaction
+   * @param baseSequence        the base sequence of the first record in the batch we are trying to append
+   * @param callback            the method to execute once the verification is either completed or returns an error
+   * @param partitionOperation  the enum value which determines whether the client is ready to handle the new Abortable Txn Errors
```
nit: to make this more generic we should say something like "determines which version of the client sent the request so the addPartitionsToTxnManager handles it appropriately".

I also wonder if we want to rename the argument. PartitionOperation is a little unclear.

jolshan commented on core/src/test/scala/unit/kafka/server/AddPartitionsToTxnManagerTest.scala:
```diff
@@ -72,6 +72,8 @@ class AddPartitionsToTxnManagerTest {
   private val versionMismatchResponse = clientResponse(null, mismatchException = new UnsupportedVersionException(""))
   private val disconnectedResponse = clientResponse(null, disconnected = true)
 
+  //private val highestProduceVersion = ApiKeys.PRODUCE.oldestVersion
```
nit -- remove commented out code

sjhajharia committed:
Renaming the Enum and minor comments

sjhajharia commented:
@jolshan, I updated the PR addressing the comments you left. I have renamed the enum to `ApiVersionErrorMapper`. Please do let me know if that looks better.
I will also wait for your confirmation on whether we need any changes in the exception naming and if all the Transactional APIs need to be bumped.
Thanks

jolshan commented:
Hey @sjhajharia thanks for the updates. I think `ApiVersionErrorMapper` is also a bit tricky since the addPartitions change will not be related to errors at all. That's why I was thinking of something a bit more generic. "Operation" was probably the right direction rather than "errors". Sorry this part is a bit nuanced. Maybe something like "supportedOperation" and the comment could be something like the supported operation based on the client's request API version?

And yes, I think we should bump all the API versions to support this new error. I can update the KIP.

jolshan commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:
```diff
@@ -3523,6 +3524,128 @@ public void testForegroundInvalidStateTransitionIsRecoverable() {
+        assertTrue(commitResult.isAcked());
+
+        assertAbortableError(AbortableTransactionException.class);
+        // make sure the exception was thrown directly from the follow-up calls.
```
nit: was this comment meant to be before further checks? I noticed this test was a little different.

sjhajharia replied:
Fixed this

jolshan commented on core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:
```diff
@@ -3040,6 +3048,7 @@ class ReplicaManagerTest {
       transactionalId = transactionalId,
       entriesPerPartition = entriesToAppend,
       responseCallback = appendCallback,
+      apiVersionErrorMapper = genericError
```
Is there a reason some of these are generic and some are default? I wonder if we can configure the test to use the "highest" enum

sjhajharia replied:
I will change all of these to use the `genericError` as the supportedOperation

jolshan replied:
That works for now. Long term, we may want another solution, but we can do in a follow up.

sjhajharia committed:
Bumping all the Transaction Apis

sjhajharia committed:
Revert unnecessary commit

sjhajharia commented:
On that note @jolshan , do we any more open questions to address or we are good to go with this PR? Or are we in flux for [this](https://github.com/apache/kafka/pull/15486#discussion_r1523597267)?

jolshan commented on core/src/test/scala/unit/kafka/server/AddPartitionsToTxnManagerTest.scala:
```diff
@@ -106,27 +106,27 @@ class AddPartitionsToTxnManagerTest {
-    addPartitionsToTxnManager.verifyTransaction(transactionalId1, producerId1, producerEpoch = 0, topicPartitions, setErrors(transaction1Errors))
-    addPartitionsToTxnManager.verifyTransaction(transactionalId2, producerId2, producerEpoch = 0, topicPartitions, setErrors(transaction2Errors))
-    addPartitionsToTxnManager.verifyTransaction(transactionalId3, producerId3, producerEpoch = 0, topicPartitions, setErrors(transaction3Errors))
+    addPartitionsToTxnManager.verifyTransaction(transactionalId1, producerId1, producerEpoch = 0, topicPartitions, setErrors(transaction1Errors), any())
```
for this test, can we use the latest produce version rather than any?

jolshan reviewed (approved):
I left one nit. I think the error name is ok for now. I sent a message to the mailing list. If we fix the nit, and don't hear from folks I can merge this. 👍 

Thanks for your hard work @sjhajharia!

sjhajharia committed:
Updated a test

sjhajharia commented:
Thank you @jolshan for the review.
I have updated the test mentioned. Hopefully it gets into trunk soon.

AndrewJSchofield commented on core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala:
```diff
@@ -42,14 +42,26 @@ object AddPartitionsToTxnManager {
+/**
+ * This is an enum which handles the Partition Response based on the Request Version and the exact operation
+ *    defaultError:       This is the default workflow which maps to cases when the Produce Request Version or the Txn_offset_commit request was lower than the first version supporting the new Error Class
+ *    genericError:       This maps to the case when the clients are updated to handle the AbortableTxnException
```
I think it's not `AbortableTxnException`. Probably ought to give it its proper name.
